### PR TITLE
Fix unstable peers and slowness in sync

### PR DESCRIPTION
**Pull request by ngotchac**
This PR fixes some issues with the stability of peer counts, and overall slowness in block sync.

Here are graphs that compares block height against time for the current `stable` and this branch:
block_heights
block_heights
Here for the peer counts:
peer_counts
peer_counts

The issue was that `sync.continue_sync` was called every time a packet was received, so it had to go through the whole list of peers, randomize it, etc.. Which could be quite expensive when connected to a lot of peers.

Now it is called on a timer, every 2.5 seconds.

### CODE REVIEW

File: ethcore/sync/src/api.rs [hunk 1]
```diff
@@ -397,7 +398,8 @@ impl NetworkProtocolHandler for SyncProtocolHandler {
 	fn initialize(&self, io: &NetworkContext) {
 		if io.subprotocol_name() != WARP_SYNC_PROTOCOL_ID {
 			io.register_timer(PEERS_TIMER, Duration::from_millis(700)).expect("Error registering peers timer");
-			io.register_timer(SYNC_TIMER, Duration::from_millis(1100)).expect("Error registering sync timer");
+			io.register_timer(MAINTAIN_SYNC_TIMER, Duration::from_millis(1100)).expect("Error registering sync timer");
+			io.register_timer(CONTINUE_SYNC_TIMER, Duration::from_millis(2_500)).expect("Error registering sync timer");
```
Comment (sorpaas) [inline ethcore/sync/src/api.rs]: Either `2500` or change the above and below line to be with `_`.

File: ethcore/sync/src/chain/mod.rs [hunk 2]
```diff
@@ -655,37 +655,35 @@ impl ChainSync {
...
+			if peers.len() > 0 {
+				trace!(
+					target: "sync",
+					"Syncing with peers: {} active, {} available, {} total",
+					self.active_peers.len(), peers.len(), self.peers.len()
+				);
 
-			for (peer_id, _) in peers {
-				self.sync_peer(io, peer_id, false);
+				random::new().shuffle(&mut peers); //TODO: sort by rating
```
Comment (sorpaas) [inline ethcore/sync/src/chain/mod.rs]: Not caused by this PR, but please create an issue and include an issue number for this TODO!

Comment (sorpaas) [inline ethcore/sync/src/chain/mod.rs]: nit: missing space `// TODO`.

Comment (ngotchac) [reply ethcore/sync/src/chain/mod.rs]: I think it's this one: https://github.com/paritytech/parity-ethereum/issues/646

File: ethcore/sync/src/chain/mod.rs [hunk 3]
```diff
@@ -655,37 +655,35 @@ impl ChainSync {
...
+			if peers.len() > 0 {
+				trace!(
+					target: "sync",
+					"Syncing with peers: {} active, {} available, {} total",
+					self.active_peers.len(), peers.len(), self.peers.len()
+				);
 
-			for (peer_id, _) in peers {
-				self.sync_peer(io, peer_id, false);
+				random::new().shuffle(&mut peers); // TODO: sort by rating
```
Comment (sorpaas) [inline ethcore/sync/src/chain/mod.rs]: @ngotchac If it's #646, would you mind to add the number here? Like this:

```suggestion
				random::new().shuffle(&mut peers); // TODO (#646): sort by rating
```

This was originally proposed for Substrate development (which now being strictly followed). Having `TODO` with only short description for many times makes it no easy to guess the original intention of the author, and thus make it difficulty for others to pick up the `TODO`. So it was decided that for PRs, all `TODO` and `FIXME` must either be removed, or include an issue number.

**Comment by andresilva** [comment]
Should we backport this as well?

**Comment by folsen** [comment]
@ngotchac just out of curiosity, how did you generate the graphs?

**Comment by ngotchac** [comment]
@folsen I wrote a small Rust script that runs the given binary with a given config basically, then prints the graphs using `gnuplot`. Are you interested? I can publish it on GH

**Comment by folsen** [comment]
@ngotchac definitely interested, seems to cross over with what some of the QA guys are working on

**Comment by ngotchac** [comment]
The script used to get the graphs : https://github.com/ngotchac/eth-metrics
It's very PoC for now.

**Pull request merged**